File: SignalServiceKit/src/Network/API/OWSUpload.m
//
//  Copyright (c) 2021 Open Whisper Systems. All rights reserved.
//

#import "OWSUpload.h"
#import <AFNetworking/AFURLRequestSerialization.h>
#import <PromiseKit/AnyPromise.h>
#import <SignalCoreKit/Cryptography.h>
#import <SignalCoreKit/NSData+OWS.h>
#import <SignalCoreKit/NSDate+OWS.h>
#import <SignalServiceKit/MIMETypeUtil.h>
#import <SignalServiceKit/OWSError.h>
#import <SignalServiceKit/OWSRequestFactory.h>
#import <SignalServiceKit/OWSSignalService.h>
#import <SignalServiceKit/SSKEnvironment.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>
#import <SignalServiceKit/TSAttachmentStream.h>
#import <SignalServiceKit/TSNetworkManager.h>
#import <SignalServiceKit/TSSocketManager.h>

NS_ASSUME_NONNULL_BEGIN

void AppendMultipartFormV4Path(id<AFMultipartFormData> formData, NSString *name, NSString *dataString)
{
    NSData *data = [dataString dataUsingEncoding:NSUTF8StringEncoding];

    [formData appendPartWithFormData:data name:name];
}

#pragma mark -

// See: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-UsingHTTPPOST.html
@implementation OWSUploadFormV4

- (instancetype)initWithPolicy:(NSString *)policy
                    credential:(NSString *)credential
                  attachmentId:(NSString *)attachmentId
{
    self = [super init];

    if (self) {
        _policy = policy;
        _credential = credential;
        _attachmentId = attachmentId;
    }
    return self;
}

+ (nullable OWSUploadFormV4 *)parseDictionary:(nullable NSDictionary *)formResponseObject
{
    if (![formResponseObject isKindOfClass:[NSDictionary class]]) {
        OWSFailDebug(@"Invalid upload form.");
        return nil;
    }
    NSDictionary *responseMap = formResponseObject;

    NSString *_Nullable formPolicy = responseMap[@"policy"];
    if (![formPolicy isKindOfClass:[NSString class]] || formPolicy.length < 1) {
        OWSFailDebug(@"Invalid upload form: acl.");
        return nil;
    }

    NSString *_Nullable formCredential = responseMap[@"credential"];
    if (![formCredential isKindOfClass:[NSString class]] || formCredential.length < 1) {
        OWSFailDebug(@"Invalid upload form: credential.");
        return nil;
    }
    
    NSString *_Nullable attachmentId = responseMap[@"attachmentId"];
    if (![attachmentId isKindOfClass:[NSString class]] || attachmentId.length < 1) {
        OWSFailDebug(@"Invalid upload form: credential.");
        return nil;
    }

    return [[OWSUploadFormV4 alloc] initWithPolicy:formPolicy
                                        credential:formCredential
                                      attachmentId:attachmentId];
}

- (void)appendToForm:(id<AFMultipartFormData>)formData
{
    AppendMultipartFormV4Path(formData, @"policy", self.policy);
}

@end

#pragma mark -

@interface OWSAvatarUploadV4 ()

@property (nonatomic, nullable) NSData *avatarData;

@end

#pragma mark -

@implementation OWSAvatarUploadV4

#pragma mark - Avatars

// If avatarData is nil, we are clearing the avatar.
- (AnyPromise *)uploadAvatarToService:(nullable NSData *)avatarData
{
    OWSAssertDebug(avatarData == nil || avatarData.length > 0);
    self.avatarData = avatarData;

    __weak OWSAvatarUploadV4 *weakSelf = self;
    AnyPromise *promise = [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            TSRequest *formRequest = [OWSRequestFactory profileAvatarUploadFormRequest];
            [self.networkManager makeRequest:formRequest
                success:^(NSURLSessionDataTask *task, id _Nullable formResponseObject) {
                    OWSAvatarUploadV4 *_Nullable strongSelf = weakSelf;
                    if (!strongSelf) {
                        return resolve(OWSErrorWithCodeDescription(OWSErrorCodeUploadFailed, @"Upload deallocated"));
                    }

                    if (avatarData == nil) {
                        OWSLogDebug(@"successfully cleared avatar");
                        return resolve(@(1));
                    }

                    [strongSelf parseFormAndUpload:formResponseObject]
                        .thenInBackground(^{ return resolve(@(1)); })
                        .catchInBackground(^(NSError *error) { resolve(error); });
                }
                failure:^(NSURLSessionDataTask *task, NSError *error) {
                    OWSLogError(@"Failed to get profile avatar upload form: %@", error);
                    resolve(error);
                }];
        });
    }];
    return promise;
}

- (AnyPromise *)parseFormAndUpload:(nullable id)formResponseObject
{
    OWSUploadFormV4 *_Nullable form = [OWSUploadFormV4 parseDictionary:formResponseObject];
    if (!form) {
        return [AnyPromise
            promiseWithValue:OWSErrorWithCodeDescription(OWSErrorCodeUploadFailed, @"Invalid upload form.")];
    }
    
    NSError *error;
    NSData *decodedData = [[NSData alloc] initWithBase64EncodedString:form.policy options:0];
    NSDictionary *decodedDictionary = [NSJSONSerialization JSONObjectWithData:decodedData options:NSJSONReadingAllowFragments error:&error];
    self.bucket = decodedDictionary[@"bucket"];
    self.urlPath = decodedDictionary[@"policy"][@"attachmentId"];
    self.credentials = decodedDictionary[@"policy"][@"credential"];
    
    NSString *uploadUrlPath = @"/api/v1/osp/objects";
    return [OWSUpload uploadV4WithData:self.avatarData
                            uploadForm:form
                         uploadUrlPath:uploadUrlPath
                         progressBlock:nil];
}

@end

//#pragma mark - Attachments
//
//@interface OWSAttachmentUploadV4 ()
//
//@property (nonatomic) TSAttachmentStream *attachmentStream;
//
//@end
//
//#pragma mark -
//
//@implementation OWSAttachmentUploadV4
//
//#pragma mark - Dependencies
//
//- (AFHTTPSessionManager *)uploadHTTPManager
//{
//    return [[OWSSignalService shared] sessionManagerForCdnNumber:0];
//}
//
//- (TSNetworkManager *)networkManager
//{
//    return SSKEnvironment.shared.networkManager;
//}
//
//- (TSSocketManager *)socketManager
//{
//    return SSKEnvironment.shared.socketManager;
//}
//
//#pragma mark -
//
//- (nullable NSData *)attachmentData
//{
//    OWSAssertDebug(self.attachmentStream);
//
//    NSData *encryptionKey;
//    NSData *digest;
//    NSError *error;
//    NSData *attachmentData = [self.attachmentStream readDataFromFileWithError:&error];
//    if (error) {
//        OWSLogError(@"Failed to read attachment data with error: %@", error);
//        return nil;
//    }
//
//    NSData *_Nullable encryptedAttachmentData = [Cryptography encryptAttachmentData:attachmentData
//                                                                          shouldPad:YES
//                                                                             outKey:&encryptionKey
//                                                                          outDigest:&digest];
//    if (!encryptedAttachmentData) {
//        OWSFailDebug(@"could not encrypt attachment data.");
//        return nil;
//    }
//
//    self.encryptionKey = encryptionKey;
//    self.digest = digest;
//
//    return encryptedAttachmentData;
//}
//
//- (AnyPromise *)uploadAttachmentToService:(TSAttachmentStream *)attachmentStream
//                            progressBlock:(UploadProgressBlock)progressBlock
//{
//    OWSAssertDebug(attachmentStream);
//
//    self.attachmentStream = attachmentStream;
//
//    AnyPromise *promise = [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
//        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
//            [self uploadAttachmentToService:resolve progressBlock:progressBlock skipWebsocket:NO];
//        });
//    }];
//    return promise;
//}
//
//- (void)uploadAttachmentToService:(PMKResolver)resolve
//                    progressBlock:(UploadProgressBlock)progressBlock
//                    skipWebsocket:(BOOL)skipWebsocket
//{
//    TSRequest *formRequest = [OWSRequestFactory allocAttachmentRequestV4];
//
//    __weak OWSAttachmentUploadV4 *weakSelf = self;
//    void (^formSuccess)(id _Nullable) = ^(id _Nullable formResponseObject) {
//            OWSAttachmentUploadV4 *_Nullable strongSelf = weakSelf;
//        if (!strongSelf) {
//            return resolve(OWSErrorWithCodeDescription(OWSErrorCodeUploadFailed, @"Upload deallocated"));
//        }
//
//        [strongSelf parseFormAndUpload:formResponseObject progressBlock:progressBlock]
//                .thenInBackground(^{ resolve(@(1)); })
//                .catchInBackground(^(NSError *error) { resolve(error); });
//    };
//    void (^formFailure)(NSError *) = ^(NSError *error) {
//        OWSLogError(@"Failed to get profile avatar upload form: %@", error);
//        resolve(error);
//    };
//
//
//    [self.networkManager makeRequest:formRequest
//        success:^(NSURLSessionDataTask *task, id _Nullable formResponseObject) {
//            formSuccess(formResponseObject);
//        }
//        failure:^(NSURLSessionDataTask *task, NSError *error) {
//            formFailure(error);
//        }
//     ];
//}
//
//#pragma mark -
//
//- (AnyPromise *)parseFormAndUpload:(nullable id)formResponseObject
//                     progressBlock:(UploadProgressBlock)progressBlock
//{
//    OWSUploadFormV4 *_Nullable form = [OWSUploadFormV4 parseDictionary:formResponseObject];
//    if (!form) {
//        return [AnyPromise
//            promiseWithValue:OWSErrorWithCodeDescription(OWSErrorCodeUploadFailed, @"Invalid upload form.")];
//    }
//
//    UInt64 serverId = [form.attachmentId longLongValue];
//    if (serverId < 1) {
//        return [AnyPromise
//            promiseWithValue:OWSErrorWithCodeDescription(OWSErrorCodeUploadFailed, @"Invalid upload form.")];
//    }
//
//    self.serverId = serverId;
//
//    NSError *error;
//    NSData *decodedData = [[NSData alloc] initWithBase64EncodedString:form.policy options:0];
//    NSDictionary *decodedDictionary = [NSJSONSerialization JSONObjectWithData:decodedData options:NSJSONReadingAllowFragments error:&error];
//    self.bucket = decodedDictionary[@"bucket"];
//    self.credentionals = decodedDictionary[@"policy"][@"credential"];
//
//    __weak OWSAttachmentUploadV4 *weakSelf = self;
//    NSString *uploadUrlPath = @"/api/v1/osp/objects";
//    return [OWSUpload uploadV4WithData:self.attachmentData
//                            uploadForm:form
//                         uploadUrlPath:uploadUrlPath
//                         progressBlock:progressBlock]
//        .then(^{
//            weakSelf.uploadTimestamp = NSDate.ows_millisecondTimeStamp;
//        });
//}
//
//@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/src/Network/API/OWSUpload.swift
//
//  Copyright (c) 2020 Open Whisper Systems. All rights reserved.
//

import Foundation
import PromiseKit

public enum OWSUploadError: Error {
    case missingRangeHeader
}

@objc
public class OWSUpload: NSObject {

    public typealias ProgressBlock = (Progress) -> Void

    @objc
    @available(swift, obsoleted: 1.0)
    public class func uploadV4(data: Data,
                               uploadForm: OWSUploadFormV4,
                               uploadUrlPath: String,
                               progressBlock: ((Progress) -> Void)?) -> AnyPromise {
        AnyPromise(uploadV4(data: data,
                            uploadForm: uploadForm,
                            uploadUrlPath: uploadUrlPath,
                            progressBlock: progressBlock))
    }

    @objc
    public static let serialQueue: DispatchQueue = {
        return DispatchQueue(label: "org.whispersystems.signal.upload",
                             qos: .utility,
                             autoreleaseFrequency: .workItem)
    }()
}

// MARK: -

fileprivate extension OWSUpload {

    static var cdn0SessionManager: AFHTTPSessionManager {
        signalService.sessionManagerForCdn(cdnNumber: 0)
    }
}

// MARK: -

public extension OWSUpload {

    class func uploadV4(data: Data,
                        uploadForm: OWSUploadFormV4,
                        uploadUrlPath: String,
                        progressBlock: ProgressBlock? = nil) -> Promise<String> {

        guard !AppExpiry.shared.isExpired else {
            return Promise(error: OWSAssertionError("App is expired."))
        }

        let (promise, resolver) = Promise<String>.pending()
        Self.serialQueue.async {
            // TODO: Use OWSUrlSession instead.
            self.cdn0SessionManager.requestSerializer.setValue(OWSURLSession.signalIosUserAgent,
                                                               forHTTPHeaderField: OWSURLSession.kUserAgentHeader)
            self.cdn0SessionManager.post(uploadUrlPath,
                                         parameters: nil,
                                         constructingBodyWith: { (formData: AFMultipartFormData) -> Void in

                                            formData.appendPart(withFileData: data, name: "file", fileName: "file", mimeType: "")
                                            
                                            // We have to build up the form manually vs. simply passing in a parameters dict
                                            // because AWS is sensitive to the order of the form params (at least the "key"
                                            // field must occur early on).
                                            //
                                            // For consistency, all fields are ordered here in a known working order.
                                            uploadForm.append(toForm: formData)
            },
                                         progress: { progress in
                                            Logger.verbose("progress: \(progress.fractionCompleted)")

                                            if let progressBlock = progressBlock {
                                                progressBlock(progress)
                                            }
            },
                                         success: { (_, _) in
                                            Logger.verbose("Success.")
                                            let uploadedUrlPath = uploadForm.attachmentId
                                            resolver.fulfill(uploadedUrlPath)
            }, failure: { (task, error) in
                if let task = task {
                    #if TESTABLE_BUILD
                    TSNetworkManager.logCurl(for: task)
                    #endif
                } else {
                    owsFailDebug("Missing task.")
                }

                if let statusCode = HTTPStatusCodeForError(error),
                statusCode.intValue == AppExpiry.appExpiredStatusCode {
                    AppExpiry.shared.setHasAppExpiredAtCurrentVersion()
                }

                owsFailDebugUnlessNetworkFailure(error)
                resolver.reject(error)
            })
        }
        return promise
    }
}

// MARK: -

public extension OWSUploadFormV4 {
    class func parse(proto attributesProto: GroupsProtoAvatarUploadAttributes) throws -> OWSUploadFormV4 {

        guard let policy = attributesProto.policy else {
            throw OWSAssertionError("Missing policy.")
        }
        guard let credential = attributesProto.credential else {
            throw OWSAssertionError("Missing credential.")
        }

        return OWSUploadFormV4(policy: policy, credential: credential, attachmentId: "")
    }
}


// MARK: -

// A strong reference should be maintained to this object
// until it completes.  If it is deallocated, the upload
// may be cancelled.
//
// This class can be safely accessed and used from any thread.
@objc
public class OWSAttachmentUploadV4: NSObject {

    public typealias ProgressBlock = (Progress) -> Void

    // These properties are only set for v2 uploads.
    // For other uploads we use these defaults.
    //
    // These properties are set on success;
    // They should not be accessed before.
    @objc
    public var encryptionKey: Data?
    @objc
    public var digest: Data?
    @objc
    public var serverId: UInt64 = 0
    @objc
    public var bucket: String?
    @objc
    public var credentionals: String?
    @objc
    public var uploadTimestamp: UInt64 = 0

    private let attachmentStream: TSAttachmentStream

    @objc
    public static var serialQueue: DispatchQueue {
        OWSUpload.serialQueue
    }

    @objc
    public required init(attachmentStream: TSAttachmentStream) {
        self.attachmentStream = attachmentStream
    }

    private func attachmentMetadata() -> Promise<(url: URL, length: Int)> {
        return firstly(on: Self.serialQueue) {
            let temporaryFile = OWSFileSystem.temporaryFileUrl(isAvailableWhileDeviceLocked: true)
            let metadata = try Cryptography.encryptAttachment(at: self.attachmentStream.originalMediaURL!, output: temporaryFile)

            self.encryptionKey = metadata.key
            self.digest = metadata.digest

            guard let length = metadata.length, let plaintextLength = metadata.plaintextLength else {
                throw OWSAssertionError("Missing length.")
            }

            guard plaintextLength <= OWSMediaUtils.kMaxFileSizeGeneric,
                  length <= OWSMediaUtils.kMaxAttachmentUploadSizeBytes else {
                throw OWSAssertionError("Data is too large: \(length).").asUnretryableError
            }

            return (temporaryFile, length)
        }
    }

    private func attachmentData() -> Promise<Data> {
        // TODO: Eliminate the need for ever loading the attachment data into memory.
        // Right now, this is only used when updating group avatars.
        return attachmentMetadata().map { try Data(contentsOf: $0.url) }
    }

    @objc
    @available(swift, obsoleted: 1.0)
    public func upload(progressBlock: ProgressBlock? = nil) -> AnyPromise {
        return AnyPromise(upload(progressBlock: progressBlock))
    }

    public func upload(progressBlock: ProgressBlock? = nil) -> Promise<Void> {
        return uploadV4(progressBlock: progressBlock)
    }

    // Performs a request, trying to use the websocket
    // and failing over to REST.
    private func performRequest(skipWebsocket: Bool = false,
                                requestBlock: @escaping () -> TSRequest) -> Promise<Any?> {
        return firstly(on: Self.serialQueue) { () -> Promise<Any?> in
            let formRequest = requestBlock()
            let shouldUseWebsocket = OWSUpload.socketManager.canMakeRequests() && !skipWebsocket
            if shouldUseWebsocket {
                return firstly(on: Self.serialQueue) { () -> Promise<Any?> in
                    OWSUpload.socketManager.makeRequestPromise(request: formRequest)
                }.recover(on: Self.serialQueue) { (_) -> Promise<Any?> in
                    // Failover to REST request.
                    self.performRequest(skipWebsocket: true, requestBlock: requestBlock)
                }
            } else {
                return firstly(on: Self.serialQueue) {
                    return OWSUpload.networkManager.makePromise(request: formRequest)
                }.map(on: Self.serialQueue) { (_: URLSessionDataTask, responseObject: Any?) -> Any? in
                    return responseObject
                }
            }
        }
    }

    // MARK: - V4

    public func uploadV4(progressBlock: ProgressBlock? = nil) -> Promise<Void> {
        return firstly(on: Self.serialQueue) {
            // Fetch attachment upload form.
            return self.performRequest {
                return OWSRequestFactory.allocAttachmentRequestV2()
            }
        }.then(on: Self.serialQueue) { [weak self] (formResponseObject: Any?) -> Promise<OWSUploadFormV4> in
            guard let self = self else {
                throw OWSAssertionError("Upload deallocated")
            }
            return self.parseUploadFormV4(formResponseObject: formResponseObject)
        }.then(on: Self.serialQueue) { (form: OWSUploadFormV4) -> Promise<(form: OWSUploadFormV4, attachmentData: Data)> in
            return firstly {
                return self.attachmentData()
            }.map(on: Self.serialQueue) { (attachmentData: Data) in
                return (form, attachmentData)
            }
        }.then(on: Self.serialQueue) { (form: OWSUploadFormV4, attachmentData: Data) -> Promise<String> in
            let uploadUrlPath = "/api/v1/osp/objects"
            return OWSUpload.uploadV4(data: attachmentData,
                                      uploadForm: form,
                                      uploadUrlPath: uploadUrlPath,
                                      progressBlock: progressBlock)
        }.map(on: Self.serialQueue) { [weak self] (_) throws -> Void in
            self?.uploadTimestamp = NSDate.ows_millisecondTimeStamp()
        }
    }

    private func parseUploadFormV4(formResponseObject: Any?) -> Promise<OWSUploadFormV4> {

        return firstly(on: Self.serialQueue) { () -> OWSUploadFormV4 in
            guard let formDictionary = formResponseObject as? [AnyHashable: Any] else {
                Logger.warn("formResponseObject: \(String(describing: formResponseObject))")
                throw OWSAssertionError("Invalid form.")
            }
            guard let form = OWSUploadFormV4.parseDictionary(formDictionary) else {
                Logger.warn("formDictionary: \(formDictionary)")
                throw OWSAssertionError("Invalid form dictionary.")
            }
            let serverId: UInt64 = UInt64(form.attachmentId) ?? 0
            guard serverId > 0 else {
                Logger.warn("serverId: \(serverId)")
                throw OWSAssertionError("Invalid serverId.")
            }

            self.serverId = serverId

            if let policyDict = Data(base64Encoded: form.policy).flatMap({ try? JSONSerialization.jsonObject(with: $0, options: .allowFragments) }) as? NSDictionary {
                self.bucket = policyDict["bucket"] as? String
                self.credentionals = (policyDict["policy"] as? NSDictionary)?["credential"] as? String
            }

            return form
        }
    }
}
